### PR TITLE
variable, main: make sure default values for sysvars validate

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Problem Summary:

Fixes https://github.com/pingcap/tidb/issues/24794 

This adds a new test to the variable package, and fixes a few inconsistencies.

If you try and set a sysvar to its default value, it should both pass validation and the value should already be normalized.

This is incredibly useful, because the most annoying thing is trying to set a setting to something like what it already is, but it fails validation.

This also increases the code-coverage for variable:
```
From:
ok  	github.com/pingcap/tidb/sessionctx/variable	0.012s	coverage: 48.7% of statements
To:
ok  	github.com/pingcap/tidb/sessionctx/variable	0.014s	coverage: 52.7% of statements
```

### What is changed and how it works?

What's Changed:

Added test and inconsistencies fixed.

### Related changes

- None

### Check List 

Tests 

- Unit test

Side effects

- None

### Release note 

- No release note

### CODE REVIEW

File: sessionctx/variable/noop.go [hunk 1]
```diff
@@ -475,7 +475,7 @@ var noopSysVars = []*SysVar{
 	// for compatibility purpose, we should leave them alone.
 	// TODO: Follow the Terminology Updates of MySQL after their changes arrived.
 	// https://mysqlhighavailability.com/mysql-terminology-updates/
-	{Scope: ScopeSession, Name: PseudoSlaveMode, Value: "", Type: TypeInt},
+	{Scope: ScopeSession, Name: PseudoSlaveMode, Value: Off, Type: TypeBool},
```
Comment (morgo) [inline sessionctx/variable/noop.go:478]: I verified this in the MySQL manual. It's supposed to be a boolean.

File: sessionctx/variable/noop.go [hunk 2]
```diff
@@ -76,7 +76,7 @@ var noopSysVars = []*SysVar{
 	{Scope: ScopeGlobal, Name: "log_backward_compatible_user_definitions", Value: ""},
 	{Scope: ScopeNone, Name: "lc_messages_dir", Value: "/usr/local/mysql-5.6.25-osx10.8-x86_64/share/"},
 	{Scope: ScopeGlobal, Name: "ft_boolean_syntax", Value: "+ -><()~*:\"\"&|"},
-	{Scope: ScopeGlobal, Name: TableDefinitionCache, Value: "-1", Type: TypeUnsigned, MinValue: 400, MaxValue: 524288, AutoConvertOutOfRange: true},
+	{Scope: ScopeGlobal, Name: TableDefinitionCache, Value: "2000", Type: TypeUnsigned, MinValue: 400, MaxValue: 524288, AutoConvertOutOfRange: true},
```
Comment (morgo) [inline sessionctx/variable/noop.go:79]: I observed 2000 from MySQL.

File: sessionctx/variable/sysvar_test.go [hunk 3]
```diff
@@ -506,12 +506,16 @@ func (*testSysVarSuite) TestInstanceScopedVars(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(val, Equals, fmt.Sprintf("%g", MemoryUsageAlarmRatio.Load()))
 
-	val, err = GetSessionOrGlobalSystemVar(vars, TiDBConfig)
-	c.Assert(err, IsNil)
-	conf := config.GetGlobalConfig()
-	j, err := json.MarshalIndent(conf, "", "\t")
-	c.Assert(err, IsNil)
-	c.Assert(val, Equals, config.HideConfig(string(j)))
+	// Because JSON key order is non-deterministic, this test is flaky.
```
Comment (xhebox) [inline sessionctx/variable/sysvar_test.go:509]: Maybe try to unmarshal the val into config?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: sessionctx/variable/sysvar_test.go (hunk 3 above)
Comment (morgo) [reply sessionctx/variable/sysvar_test.go]: I figured out what causes it. It was not json key order, but a test in `server/http_handler_test.go` which changed the expected value and did not restore it back.

File: server/http_handler_test.go [hunk 4]
```diff
@@ -1228,6 +1228,8 @@ func (ts *HTTPHandlerTestSuite) TestPostSettings(c *C) {
 	c.Assert(resp.StatusCode, Equals, http.StatusOK)
 	c.Assert(config.GetGlobalConfig().CheckMb4ValueInUTF8, Equals, false)
 	dbt.mustExec("insert t2 values (unhex('f09f8c80'));")
+	// restore original value.
+	config.GetGlobalConfig().CheckMb4ValueInUTF8 = true
```
Comment (xhebox) [inline server/http_handler_test.go:1231]: I think test cases modifying global config better go to serial tests. As it may cause data race.

**Review by xhebox — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]

**Pull request merged**